crates/typos-cli/src/bin/typos-cli/report.rs

```diff
@@ -280,6 +287,132 @@ impl Report for PrintJson {
     }
 }
 
+#[derive(Debug)]
+pub(crate) struct PrintSarif {
+    results: Mutex<RefCell<Vec<sarif::Result>>>,
+    error: Mutex<RefCell<Option<String>>>,
+}
+
+impl Default for PrintSarif {
+    fn default() -> Self {
+        Self {
+            results: Mutex::new(RefCell::new(Vec::new())),
+            error: Mutex::new(RefCell::new(None)),
+        }
+    }
+}
+
+impl Report for PrintSarif {
+    fn report(&self, msg: Message<'_>) -> Result<(), std::io::Error> {
+        match &msg {
+            Message::BinaryFile(_) => {
+                // todo: implement artifacts for sarif
+            }
+            Message::FileType(_) => {
+                // todo: implement artifacts for sarif
+            }
+            Message::File(_) => {
+                // todo: implement artifacts for sarif
+            }
+            Message::Typo(msg) => {
+                let start = String::from_utf8_lossy(&msg.buffer[0..msg.byte_offset]);
+                let column_number =
+                    unicode_segmentation::UnicodeSegmentation::graphemes(start.as_ref(), true).count() + 1;
+                let line_num = msg.context.as_ref().map_or(1, |context| match context {
+                    Context::File(context) => context.line_num,
+                    _ => 1,
+                });
+
+                let message = match &msg.corrections {
+                    typos::Status::Valid => { return Ok(()) }
+                    typos::Status::Invalid => {
+                        format!("`{}` is disallowed", msg.typo)
+                    }
+                    typos::Status::Corrections(corrections) => {
+                        format!("`{}` should be {}", msg.typo, itertools::join(
+                            corrections.iter().map(|s| format!("`{}`", s)),
+                            ", ",
+                        ))
+                    }
+                };
```

Comment:
nit: seems like this should go first to avoid doing stuff in the presence of an early-return

Reply:
I don't know how to do this more `rust`. even I check it earlier, the match statement still needs a branch for that. use unreachable!？

Reply:
If its truly unreachable, sure, but that should be verified and likely should have a message with it to say why you know its unreachable.

If it isn't truely unreachable, then please move this block before `let start`.

Reply:
I don't think this is resolved.

Reply:
ok